### ISSUE TITLE
Original context dropped in prpc decoding

### ISSUE THREAD

**Issue by code4ghana** [luci/luci-go#29]
Currently, the original incoming context is dropped in the decoder.parseHeader function.
As a result it's not possible to make calls to most appengine require the context.
The context passed by prpc is different and does not appear to satisfy the libs. work for such calls.

For example urlFetch calls fail with "not an App Engine context".

```
import (
    "google.golang.org/appengine/urlfetch"
)
func (s* funkyService) getFunky(c context.Context, req *funky.GetFunkyRequest) (...) {
    client := appengine.urlFetch.Client(context.Context)
    r, err := client.Get("http://google.com/")
}
```

Furthermore, I  can't call  appengine.NewContext(*http.request) since the http.Request object is not available.

any thoughts?

**Comment by sergeyberezin** [comment]
It's probably the same as https://github.com/luci/gae/issues/63 

TL;DR:
```Go
import "github.com/luci/gae/impl/prod"
...
client := appenigne.urlFetch.Client(prod.AEContext(c.Context))
```

**Comment by code4ghana** [comment]
Thanks for the response,
I tried your suggestion in my local dev environment but it's not working I've pasted the trace below.

I believe that you were referring to the router.Context passed to Middlewares. The rpc handler gets passed context.Context , I'm not sure if this is the same as the context inside router.Context.


```
func (s *authenticationService) VerifyToken(c context.Context, req *authentication.VerifyTokenRequest) (*authentication.VerifyTokenResponse, error) {
	log.Printf("***CONTEXT: %v/r", c)
	ctx := prod.AEContext(c)
```
Start the server with `$goapp serve server`

Stack trace:
```
**CONTEXT: context.Background.WithValue(metadata.mdKey{}, metadata.MD{"X-Appengine-Server-Software":[]string{"Development/2.0"}, "User-Agent":[]string{"curl/7.43.0"}, "X-Appengine-User-Nickname":[]string{""}, "X-Appengine-Server-Protocol":[]string{"HTTP/1.1"}, "X-Appengine-User-Email":[]string{""}, "Content-Length":[]string{"212"}, "X-Appengine-Server-Name":[]string{""}, "Accept-Encoding":[]string{"identity"}, "X-Appengine-User-Organization":[]string{""}, "X-Appengine-Country":[]string{"ZZ"}, "X-Appengine-Server-Port":[]string{"8080"}, "X-Appengine-Request-Id-Hash":[]string{"7EFC56F0"}, "X-Appengine-User-Id":[]string{""}, "X-Appengine-Default-Version-Hostname":[]string{":8080"}, "X-Appengine-User-Is-Admin":[]string{"0"}, "X-Appengine-Remote-Addr":[]string{"::1"}, "X-Appengine-Request-Log-Id":[]string{"be872b923d206ca1e01b20ee8babfe5aeadeecbaabde1da"}})/r

2016/10/27 08:21:23 http: panic serving 127.0.0.1:55867: interface conversion: interface {} is nil, not *prod.prodState
goroutine 34 [running]:
net/http.(*conn).serve.func1(0xc82008a100)
	/go_appengine/goroot/src/net/http/server.go:1389 +0xc1
panic(0x6f3800, 0xc820220680)
	/go_appengine/goroot/src/runtime/panic.go:443 +0x4e9
github.com/luci/gae/impl/prod.getProdState(0xfaf2d0, 0xc82021f020, 0x0, 0x0, 0x0, 0x0, 0x0)
	/workspace/golang/src/github.com/luci/gae/impl/prod/context.go:158 +0x103
github.com/luci/gae/impl/prod.AEContext(0xfaf2d0, 0xc82021f020, 0x0, 0x0)
	/workspace/golang/src/github.com/luci/gae/impl/prod/context.go:41 +0x3e

main58274.(*authenticationService).VerifyToken(0xc57110, 0xfaf2d0, 0xc82021f020, 0xc8204883e0, 0xc82021ef30, 0x0, 0x0)
	service.go:42 +0x23b
  myrepo.com/authentication/proto._AuthenticationService_VerifyToken_Handler(0x6b95c0, 0xc57110, 0xfaf2d0, 0xc82021f020, 0xc820480330, 0x0, 0x0, 0x0, 0x0, 0x0)
	/workspace/golang/src/bitbucket.com/code4ghana/authentication/proto/service.pb.go:128 +0x168

github.com/luci/luci-go/grpc/prpc.(*method).handle(0xc82026a1c0, 0xfaf2d0, 0xc82021f020, 0xfaf200, 0xc8202b6000, 0xc8202102a0, 0x0, 0x0)
	/workspace/golang/src/github.com/luci/luci-go/grpc/prpc/method.go:53 +0x258
github.com/luci/luci-go/grpc/prpc.(*Server).respond(0xc8202208c0, 0xe0f788, 0xc8200712f0, 0xfaf200, 0xc8202b6000, 0xc8202102a0, 0xc82021c28b, 0x24, 0xc82021c2b0, 0xb, ...)
	/workspace/golang/src/github.com/luci/luci-go/grpc/prpc/server.go:201 +0x419
github.com/luci/luci-go/grpc/prpc.(*Server).handlePOST(0xc8202208c0, 0xc820220280)
	/workspace/golang/src/github.com/luci/luci-go/grpc/prpc/server.go:166 +0x146
github.com/luci/luci-go/grpc/prpc.(*Server).(github.com/luci/luci-go/grpc/prpc.handlePOST)-fm(0xc820220280)
	/golang/src/github.com/luci/luci-go/grpc/prpc/server.go:156 +0x2a
github.com/luci/luci-go/server/router.run(0xc820220280, 0xc82048c688, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc82048c690)
	/golang/src/github.com/luci/luci-go/server/router/handler.go:85 +0x383
github.com/luci/luci-go/server/router.run(0xc820220280, 0xc82048c688, 0x1, 0x1, 0x0, 0x0, 0x0, 0xc82048c690)
	/golang/src/github.com/luci/luci-go/server/router/handler.go:76 +0x1eb
github.com/luci/luci-go/server/router.run.func1(0xc820220280)
	/golang/src/github.com/luci/luci-go/server/router/handler.go:74 +0x9c
```

**Comment by code4ghana** [comment]
I also tried your suggestion in a router.Hanlder and still got the same error.

**Comment by danjacques** [comment]
There are two different Contexts at work here, unfortunately.

First is the Context that you have constructed for your application, which should have been built with `impl/prod`'s `Use` method. This installs the `luci/gae` services into your `Context` and makes it candidate for `luci/gae` method calling.

The second is the actual AppEngine `Context`, which must be supplied when calling AppEngine methods directly (e.g., `urlfetch`). You extract this from the `luci/gae` `Context` using the `AEContext` method.

The panic that you are seeing is due to `AEContext` on a `Context` that does **not** have `luci/gae`'s services installed.

However, if your goal is to use the `urlfetch` service, consider not juggling `Context` around and just using `luci/gae`'s [urlfetch](https://godoc.org/github.com/luci/gae/service/urlfetch) service directly. Because it is a `luci/gae` service (instead of an AppEngine SDK package method), your `impl/prod`-derived `Context` should work without any funky stuff.

Unfortunately, `AEContext` is an ugly implementation detail. In general, its use is discouraged in favor of using the `luci/gae` service equivalent methods.

**Comment by nodirt** [comment]
@code4ghana did you resolve your issue after our conversation?

**Comment by code4ghana** [comment]
Hi @nodirt,

Your advice resolved it. The issue was that I was not using [gaemidleware](https://github.com/luci/luci-go/blob/master/appengine/gaemiddleware/context.go#L105). This meant that the Appengine context was getting lost.  Installing that middleware resolved it.

thanks for your patience and help!

**Comment by nodirt** [comment]
Happy to help
